File: 284. Peeking Iterator.java
//注意peek的逻辑要先判断队列中有没有，而且不一定要使用队列。
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html
class PeekingIterator implements Iterator<Integer>
{
    static Queue<Integer> que;
    static Iterator<Integer> it;
     public PeekingIterator(Iterator<Integer> iterator) {
         // initialize any member here.
         que = new LinkedList<Integer>();
         it = iterator;
     }

    // Returns the next element in the iteration without advancing the iterator.
     public Integer peek() {
         if(!que.isEmpty())
             return que.peek();
         else
         {
            int n = it.next();
            que.offer(n);
            return n;
         }
     }

     // hasNext() and next() should behave the same as in the Iterator interface.
     // Override them if needed.
     @Override
     public Integer next() {
         if(!que.isEmpty())
             return que.poll();
         else
             return it.next();
     }

     @Override
     public boolean hasNext()
     {
         if(!que.isEmpty())
             return true;
         else
             return it.hasNext();  
     }
}

File: 284. Peeking Iterator.cpp
// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.
class Iterator {
    struct Data;
	Data* data;
public:
	Iterator(const vector<int>& nums);
	Iterator(const Iterator& iter);
	virtual ~Iterator();
	// Returns the next element in the iteration.
	int next();
	// Returns true if the iteration has more elements.
	bool hasNext() const;
};


class PeekingIterator : public Iterator {
public:
	PeekingIterator(const vector<int>& nums) : Iterator(nums) {
	    // Initialize any member here.
	    // **DO NOT** save a copy of nums and manipulate it directly.
	    // You should only use the Iterator interface methods.
	    
	}

    // Returns the next element in the iteration without advancing the iterator.
	int peek() {
        return Iterator(*this).next();	//拷贝构造函数，每一次都拷贝一个副本做peek，浪费空间。
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next() {
	    return Iterator::next();
	}

	bool hasNext() const {
	    return Iterator::hasNext();
	}
};